immer/single_hash_index.hpp

```diff
@@ -0,0 +1,318 @@
+#pragma once
+
+#include <immer/config.hpp>
+#include <immer/detail/hamts/champ.hpp>
+#include <immer/detail/hamts/champ_iterator.hpp>
+#include <immer/memory_policy.hpp>
+#include <type_traits>
+
+namespace immer {
+
+template <typename T,
+          typename KeyExtractor,
+          typename Hash,
+          typename Equal,
+          typename MemoryPolicy,
+          detail::hamts::bits_t B>
+class single_hash_index_transient;
+
+template <class T>
+struct default_key_extractor
+{
+    using key_type = decltype(T().id);
+
+    key_type operator()(const T& t) const { return t.id; }
+};
```

Comment:
I would suggest that the `default_key_extractor` delegate in a free function. That way the free function can be used as customization point in a way simpler than providing a custom type. This also allows us to provide multiple default syntaxes supported. This could look like this:
```c++
template <typename T>
auto get_key(const T& x) -> decltype(x.id) { return x.id; }

struct key_fn
{
    template <typename T>
    decltype(auto) operator()(const T& x) const { return get_key(x); }
};
```

Note also how this simplifies the definition of the extractor, not needing to make the whole type a template. In order to define `key_type` in the `table` itself, you can do:
```c++
using K = std::decay_t<decltype(KeyFn{}(std::declval<T>()))>;
```